### PR TITLE
Add some paranoia

### PR DESCRIPTION
**Pull request by michaelcahill**
This is a diagnostic branch: http://build.wiredtiger.com:8080/job/wiredtiger-test-format-stress-sanitizer has been switched over to it to try to catch some hard-to-reproduce bugs.

### CODE REVIEW

**Comment by keithbostic** [comment]
@michaelcahill, just FYI, I've got 6 instances of test/format on the paranoia branch on an c3.4xlarge AWS instance, using the two CONFIGs from #1582, nothing yet.

File: src/btree/row_key.c [hunk 1]
```diff
@@ -496,6 +499,12 @@ __wt_row_ikey(WT_SESSION_IMPL *session,
...
 	memcpy(WT_IKEY_DATA(ikey), key, size);
 
-	*(WT_IKEY **)ikeyp = ikey;
+	oldv = (uintptr_t)*ikeyp;
+	WT_DIAGNOSTIC_YIELD;
+
+	WT_ASSERT(session, oldv == 0 || (oldv & WT_IK_FLAG) != 0);
+
+	if (!WT_ATOMIC_CAS8(*ikeyp, (WT_IKEY *)oldv, ikey))
+		__wt_panic(session);
```
Comment (michaelcahill) [inline src/btree/row_key.c]: We need to handle an error here properly: not just panic without any message.

File: src/conn/conn_dhandle.c [hunk 2]
```diff
@@ -730,7 +732,10 @@ __wt_conn_dhandle_discard_single(WT_SESSION_IMPL *session, int final)
 
 	if (F_ISSET(dhandle, WT_DHANDLE_OPEN)) {
 		ret = __wt_conn_btree_sync_and_close(session, 0);
-		if (!final)
+		if (final && ret != 0)
+			__wt_err(session, ret,
+			    "Final close of %s failed", dhandle->name);
```
Comment (michaelcahill) [inline src/conn/conn_dhandle.c]: I'm open to suggestions about what to do here: previously, we silently ignored failures...

File: src/btree/bt_split.c [hunk 3]
```diff
@@ -586,6 +586,16 @@ __split_deepen(WT_SESSION_IMPL *session, WT_PAGE *parent, uint32_t children)
 	 * fails, we don't roll back that change, because threads may already
 	 * be using the new index.
 	 */
+	for (parent_refp = pindex->index, i = 0;
```
Comment (agorrod) [inline src/btree/bt_split.c]: I understand that we are inside a split so it's probably safe, but I'd be inclined to use ```WT_INTL_FOREACH_BEGIN``` style loop here. Is there a reason not to?

Comment (michaelcahill) [inline src/btree/bt_split.c]: Hi @agorrod, that macro walks the index that is currently attached to an internal page.  At this point in the split code, we have already swapped in a new index, and we are walking the old one here to schedule any instantiated keys to be freed.

File: src/btree/row_key.c [hunk 4]
```diff
@@ -496,6 +500,10 @@ __wt_row_ikey(WT_SESSION_IMPL *session,
 	ikey->cell_offset = cell_offset;
 	memcpy(WT_IKEY_DATA(ikey), key, size);
 
-	*(WT_IKEY **)ikeyp = ikey;
+	oldv = (uintptr_t)*ikeyp;
+	WT_DIAGNOSTIC_YIELD;
+	WT_ASSERT(session, oldv == 0 || (oldv & WT_IK_FLAG) != 0);
+	success = WT_ATOMIC_CAS8(*ikeyp, (WT_IKEY *)oldv, ikey);
+	WT_ASSERT(session, success);
```
Comment (agorrod) [inline src/btree/row_key.c]: If the CAS and swap fails, we'll end up returning 0 and not switching in the new key. Should we have completely different code paths here, where the non-diagnostic version works like the old version?

File: src/conn/conn_dhandle.c [hunk 5]
```diff
@@ -725,13 +727,18 @@ __wt_conn_dhandle_discard_single(WT_SESSION_IMPL *session, int final)
...
 	if (F_ISSET(dhandle, WT_DHANDLE_OPEN)) {
-		ret = __wt_conn_btree_sync_and_close(session, 0);
-		if (!final)
-			WT_RET(ret);
+		tret = __wt_conn_btree_sync_and_close(session, 0);
+		if (final && tret != 0) {
+			__wt_err(session, tret,
+			    "Final close of %s failed", dhandle->name);
+			WT_TRET(tret);
+		} else
```
Comment (agorrod) [inline src/conn/conn_dhandle.c]: I think this should be ```else if (!final)```. That is we always want to return the error if this isn't a final discard.

**Comment by agorrod** [comment]
@michaelcahill One more comment above, then I'm happy for this to be merged.

**Comment by agorrod** [comment]
@michaelcahill I pushed the change I referenced. This branch looks good to me now. Feel free to merge if you are happy.

**Pull request merged**